src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb

```diff
@@ -992,5 +992,386 @@ IForLoopStatement (LoopKind.For) (OperationKind.LoopStatement) (Syntax: 'For A =
+        Expression: ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'y')
+            Left: ISyntheticLocalReferenceExpression (SynthesizedLocalKind.ForLoopLimitValue) (OperationKind.SyntheticLocalReferenceExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'y')
+            Right: IConversionExpression (ConversionKind.Basic, Implicit) (OperationKind.ConversionExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'y')
+                Operand: ILocalReferenceExpression: y (OperationKind.LocalReferenceExpression, Type: System.Int32) (Syntax: 'y')
+      IExpressionStatement (OperationKind.ExpressionStatement) (Syntax: 'For x = 12  ... Next')
+        Expression: ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'For x = 12  ... Next')
+            Left: ISyntheticLocalReferenceExpression (SynthesizedLocalKind.ForLoopStepValue) (OperationKind.SyntheticLocalReferenceExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'For x = 12  ... Next')
+            Right: IConversionExpression (ConversionKind.Basic, Explicit) (OperationKind.ConversionExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'For x = 12  ... Next')
+                Operand: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: 'For x = 12  ... Next')
+  AtLoopBottom:
+      IExpressionStatement (OperationKind.ExpressionStatement) (Syntax: 'For x = 12  ... Next')
+        Expression: ICompoundAssignmentExpression (BinaryOperationKind.IntegerAdd-IsLifted) (OperationKind.CompoundAssignmentExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'For x = 12  ... Next')
+            Left: ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'x')
+            Right: ISyntheticLocalReferenceExpression (SynthesizedLocalKind.ForLoopStepValue) (OperationKind.SyntheticLocalReferenceExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'For x = 12  ... Next')
+  Body: IBlockStatement (0 statements) (OperationKind.BlockStatement) (Syntax: 'For x = 12  ... Next')]]>.Value
+
+            VerifyOperationTreeForTest(Of ForBlockSyntax)(source, expectedOperationTree)
+        End Sub
+
+        <Fact>
+        Public Sub VerifyForToLoop3()
+            Dim source = <![CDATA[
+Structure C
+    Sub Foo()
+        Dim x As Integer
+        Dim y As Integer? = 16
+        For x = 12 To y 'BIND:"For x = 12 To y"
+        Next
+    End Sub
+End Structure
+]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IForLoopStatement (LoopKind.For) (OperationKind.LoopStatement) (Syntax: 'For x = 12  ... Next')
+  Condition: IBinaryOperatorExpression (BinaryOperationKind.IntegerLessThanOrEqual) (OperationKind.BinaryOperatorExpression, Type: System.Boolean) (Syntax: 'y')
```

Comment:
The compiler is somewhat interesting here... the limit value is actually viewed to have a 'narrowing nullable' conversion from "int?" to "int" here.  

I've viewed the IL generated here and that matches what's going on:

```
IL_0004: call instance void valuetype [mscorlib]System.Nullable`1<int32>::.ctor(!0)
	IL_0009: ldloca.s y
	IL_000b: call instance !0 valuetype [mscorlib]System.Nullable`1<int32>::get_Value()
	IL_0010: stloc.2
	IL_0011: ldc.i4.s 12
	IL_0013: stloc.0
	IL_0014: br.s IL_001a
	// loop start (head: IL_001a)
		IL_0016: ldloc.0
		IL_0017: ldc.i4.1
		IL_0018: add.ovf
		IL_0019: stloc.0

		IL_001a: ldloc.0
		IL_001b: ldloc.2
		IL_001c: ble.s IL_0016
	// end loop
```

In other words, afaict, the way this works is that instead of comparing against the boxed nullable value as we go through the loop, the boxed nullable is unboxed to get its value (which will throw if it is null), and then we compare against that as we go through.  So this seems by design right now.